### PR TITLE
[stdlib] Improve default implementation of Collection.distance(from:to:)

### PR DESCRIPTION
**Pull request by moiseev**

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please test

**Review by natecook1000 — APPROVED** [review]
LGTM!

**Comment by moiseev** [comment]
@swift-ci Please clean test

File: stdlib/public/core/Collection.swift [hunk 1]
```diff
@@ -965,30 +960,34 @@ extension Collection {
...
   @_inlineable
   public func distance(from start: Index, to end: Index) -> IndexDistance {
-    _precondition(start <= end,
-      "Only BidirectionalCollections can have end come before start")
-
-    var start = start
+    var _start: Index
+    let _end: Index
+    let step: IndexDistance
+    if start > end {
```
Comment (jrose-apple) [inline stdlib/public/core/Collection.swift]: You can use `>` on indexes?

Comment (natecook1000) [reply stdlib/public/core/Collection.swift]: Yep! `Comparable` is their one requirement.

Comment (jrose-apple) [reply stdlib/public/core/Collection.swift]: Huh, somehow I had it in my mind that they were just Equatable. Well this is nice, then.

Comment (natecook1000) [reply stdlib/public/core/Collection.swift]: You're not wrong! That was the case before the big collections-move-indices change: http://swiftdoc.org/v2.2/protocol/ForwardIndexType/

**Comment by moiseev** [comment]
Reverted in https://github.com/apple/swift/pull/13328

**Pull request merged**